File: 3. Linked List/13_add_lists.java
class Node<T> {
    T val;
    Node<T> next;

    public Node(T val) {
        this.val = val;
        this.next = null;
    }
}

class Source {

    public static Node<Integer> addLists(Node<Integer> head1, Node<Integer> head2) {
        return addListsRecursive(head1, head2, 0);
    }

    public static Node<Integer> addListsRecursive(Node<Integer> head1, Node<Integer> head2, int carry) {
        if (head1 == null && head2 == null && carry == 0) {
            return null;
        }

        int value1 = head1 == null ? 0 : head1.val;
        int value2 = head2 == null ? 0 : head2.val;

        int sum = value1 + value2 + carry;
        int digit = sum % 10;

        Node<Integer> newNode = new Node<Integer>(digit);

        int nextCarry = sum >= 10 ? 1 : 0;

        Node<Integer> next1 = head1 == null ? null : head1.next;
        Node<Integer> next2 = head2 == null ? null : head2.next;

        newNode.next = addListsRecursive(next1, next2, nextCarry);

        return newNode;
    }
}

File: 3. Linked List/13_add_lists/13_add_lists.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def add_lists(head1, head2):
    return add_lists_recursive(head1, head2, 0)

def add_lists_recursive(head1, head2, carry):
    if head1 is None and head2 is None and carry == 0:
        return None

    value1 = 0 if head1 is None else head1.val
    value2 = 0 if head2 is None else head2.val

    sum_value = value1 + value2 + carry
    digit = sum_value % 10

    new_node = Node(digit)

    next_carry = 1 if sum_value >= 10 else 0

    next1 = None if head1 is None else head1.next
    next2 = None if head2 is None else head2.next

    new_node.next = add_lists_recursive(next1, next2, next_carry)

    return new_node
